README.md
instaR: Access to Instagram API via R
---------

This package provides a series of functions that allow R users to access Instagram's API to search for media that mention specific hashtags or were sent from a certain location, to download recent pictures from a given users, and to capture lists of followers.

__IMPORTANT__: according to the [updated platform policy](https://www.instagram.com/developer/changelog/), access to most endpoints of the Instagram API (including those that return public data) requires previous approval by Instagram. More information about the permission review process can be found [here](https://www.instagram.com/developer/review/). 

## Getting Started from GitHub
To install from github, first install the devtools package and and then run ``devtools::install_github("pablobarbera/instaR/instaR")``

The httpuv package is needed to trigger the web request when authorizing against Instragram using instaOAuth. When configuring the client on https://instagram.com/developer/clients/ to receive your `app_id` and `app_secret` ensure the REDIRECT URI includes the trailing / e.g. `http://localhost:1410/`

Functionalities include:
* Downloading pictures from users / using hashtags
* Getting comments for individual posts
* Getting likes for individual posts
* Getting the most popular posts
* Getting basic user information
* Getting a hash tag count

For examples please see examples.R

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-1191254-10']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>



instaR/ChangeLog
2014-02-01 - version 0.1:
   - Initial release on GitHub
2014-02-20 - version 0.1.1:
    - Added getTagCount function
2014-03-20 - version 0.1.2
    - Now compatible with httr 0.3
2015-06-15 - version 0.1.3
	- added getLikes function
2015-08-06 - version 0.1.4
	- fixes bug in instaOAuth
2015-10-25 - version 0.1.5
	- getPopular function
	- minor fixes
2015-10-31 - version 0.1.6
	- getLocation function
2015-11-03 - version 0.2.0
	- minor fixes
2015-11-04 - version 0.2.1
	- minor fixes
2015-11-09 - version 0.2.2
	- first CRAN release
2016-07-13 - version 0.2.3
	- more informative errors messages re: API changes
	- now depends on jsonlite instead of rjson
2016-07-13 - version 0.2.3
	- more informative errors messages re: API changes
	- now depends on jsonlite instead of rjson
2017-06-01 - version 0.2.5
	- minor fixes


instaR/R/instaOAuth.R
#' @rdname instaOAuth
#' @export
#'
#' @title 
#' Create OAuth token to Intagram R session
#'
#' @description
#' \code{instaOAuth} creates an OAuth access token that enables R to make
#' authenticated calls to the Instagram API. The token can be saved as a
#' file in disk to be re-used in future sessions. This function relies on the
#' \code{httr} package to create the OAuth token, and is a simplified version
#' of one of its examples.
#'
#' @details
#' To obtain your \code{app_id} and \code{app_secret}, do the following steps.
#' First, go to \url{http://instagram.com/developer/} and register your
#' application with any name. Then, run the \code{instaOAuth} function with
#' your "Client ID" and "Client Secret" as arguments. It will return a URL,
#' which you will need to paste into the field "OAuth redirect_uri" in
#' your application settings on Instagram. After changing it, press Enter in
#' R. A new browser window will open and sign the token. If everything works
#' works well, you will get a message that says you can return to R.
#'
#'
#' @author
#' Pablo Barbera \email{pablo.barbera@@nyu.edu}
#' @seealso \code{\link{searchInstagram}}
#'
#' @param app_id numeric, Client ID of application to be used to create OAUth token. Available
#' at \url{http://instagram.com/developer}
#' 
#' @param app_secret string, Client Secret of application to be used to create OAUth token.
#' Available at \url{http://instagram.com/developer}.
#'
#' @param scope string, specifies scope of access to the authenticated user data. See
#' \url{http://instagram.com/developer/authentication/#scope} for available options.
#'
#' @examples \dontrun{
#' ## an example of an authenticated request after creating the OAuth token
#' ## where app_id and app_secret are fictitious, and token is saved for
#' ## future sessions
#'  my_oauth <- instaOAuth(app_id="123456789", app_secret="1A2B3C4D")
#'  save(my_oauth, file="my_oauth")
#'  load("my_oauth")
#'  obama <- searchInstagram(tag="obama", token=my_oauth)
#' }
#'

instaOAuth <- function(app_id, app_secret, scope=c("basic", "public_content")){

    ## getting callback URL
    full_url <- httr::oauth_callback()
    full_url <- gsub("(.*localhost:[0-9]{1,5}/).*", x=full_url, replacement="\\1")
    ## fixing most common error
    if (full_url=="http://localhost:1410") full_url <- 'http://localhost:1410/'

    message <- paste("Copy and paste into 'OAuth redirect_uri' on Instagram App Settings:", 
        full_url, "\nWhen done, press any key to continue...")
    ## prompting user to introduce callback URL in app page
    invisible(readline(message))
    ## using httr package functions
    instagram <- httr::oauth_endpoint(NULL, "authorize", "access_token",
        base_url = "https://api.instagram.com/oauth")
    myapp <- httr::oauth_app("instagram", app_id, app_secret)

    ## before httr 0.3
    if (packageVersion('httr') <= "0.2"){
        insta_token <- httr::oauth2.0_token(instagram, myapp, scope=scope)
        token <- httr::sign_oauth2.0(insta_token$access_token)
        if (httr::GET("https://api.instagram.com/v1/users/self/?", 
            config=token)$status==200){
            message("Authentication successful.")
        }
    }

    ## httr 0.3 to 0.6.1
    if (packageVersion('httr') > "0.2" & packageVersion('httr') <= "0.6.1"){
        token <- httr::oauth2.0_token(instagram, myapp, cache=FALSE, scope=scope)
        if (httr::GET(paste0("https://api.instagram.com/v1/users/self/?", 
                "&access_token=", 
                token$credentials$access_token))$status==200){
            message("Authentication successful.")
        }   
    }
    
    ## httr version from 0.6 to 1.1
    if (packageVersion('httr') > "0.6.1" & packageVersion('httr') < "1.2"){
        Sys.setenv("HTTR_SERVER_PORT" = "1410/")
        token <- httr::oauth2.0_token(instagram, myapp, cache=FALSE, scope=scope)
        if (httr::GET(paste0("https://api.instagram.com/v1/users/self/?", 
                "&access_token=", 
                token$credentials$access_token))$status==200){
            message("Authentication successful.")
        }  
    }

    ## current version of httr
    if (packageVersion('httr') >= "1.2"){
        token <- oauth2.0_token(instagram, myapp, cache=FALSE, scope=scope)
        if (GET(paste0("https://api.instagram.com/v1/users/self/?", 
                "&access_token=", 
                token$credentials$access_token))$status==200){
            message("Authentication successful.")
        }  
    }
    return(token)
}










